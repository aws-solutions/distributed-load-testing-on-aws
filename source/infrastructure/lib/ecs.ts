// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource, Construct, RemovalPolicy } from '@aws-cdk/core';
import { Repository } from '@aws-cdk/aws-ecr';
import { CfnCluster, CfnTaskDefinition } from '@aws-cdk/aws-ecs';
import { ManagedPolicy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from '@aws-cdk/aws-ec2';

/**
 * FargateECSTestRunnerContruct props
 * @interface FargateECSTestRunnerContructProps
 */
export interface FargateECSTestRunnerContructProps {
    // Fargate VPC ID
    readonly DLTfargateVpcId: string;
    // IP CIDR for Fargate egress
    readonly securityGroupEgress: string;
    // Solution ID
    readonly solutionId: string;
}

/**
 * @class
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class FargateECSTestRunnerContruct extends Construct {
    public dltEcsClusterName: string;
    public dltCloudWatchLogGroup: LogGroup;
    public dltTaskDefinitionArn: string;
    public dltTaskExecutionRole: Role;
    public dltSecurityGroupId: string;

    constructor(scope: Construct, id: string, props: FargateECSTestRunnerContructProps) {
        super(scope, id);

        const dltEcr = new Repository(this, 'DLTECR', {
            imageScanOnPush: true
        });
        dltEcr.applyRemovalPolicy(RemovalPolicy.RETAIN);

        const dltEcsCluster = new CfnCluster(this, 'DLTEcsCluster', {
            clusterName: Aws.STACK_NAME,
            clusterSettings: [{ 'name': 'containerInsights', 'value': 'enabled' }],
            tags: [
                {
                    'key': 'SolutionId',
                    'value': props.solutionId
                },
                {
                    'key': 'CloudFormation Stack',
                    'value': Aws.STACK_NAME
                }
            ]

        });

        this.dltEcsClusterName = dltEcsCluster.ref;

        this.dltTaskExecutionRole = new Role(this, 'DLTTaskExecutionRole', {
            assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
            managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonECSTaskExecutionRolePolicy')]
        });

        this.dltCloudWatchLogGroup = new LogGroup(this, 'DLTCloudWatchLogsGroup', {
            retention: RetentionDays.ONE_YEAR
        });
        const dltLogsGroupResource = this.dltCloudWatchLogGroup.node.defaultChild as CfnResource;
        dltLogsGroupResource.addMetadata('cfn_nag', {
            rules_to_suppress: [{
                id: 'W84',
                reason: 'KMS encryption unnecessary for log group'
            }]
        });

        const dltTaskDefinition = new CfnTaskDefinition(this, 'DLTTaskDefinition', {
            cpu: '2048',
            memory: '4096',
            networkMode: 'awsvpc',
            executionRoleArn: this.dltTaskExecutionRole.roleArn,
            requiresCompatibilities: ['FARGATE'],
            taskRoleArn: this.dltTaskExecutionRole.roleArn,
            containerDefinitions: [
                {
                    essential: true,
                    name: `${Aws.STACK_NAME}-load-tester`,
                    image: 'PUBLIC_ECR_REGISTRY:PUBLIC_ECR_TAG',
                    memory: 4096,
                    logConfiguration: {
                        logDriver: 'awslogs',
                        options: {
                            'awslogs-group': this.dltCloudWatchLogGroup.logGroupName,
                            'awslogs-stream-prefix': 'load-testing',
                            'awslogs-region': `${Aws.REGION}`
                        }
                    }
                }
            ],
        });

        this.dltTaskDefinitionArn = dltTaskDefinition.ref;

        const dltEcsSecurityGroup = new CfnSecurityGroup(this, 'DLTEcsSecurityGroup', {
            vpcId: props.DLTfargateVpcId,
            groupDescription: 'DLTS Tasks Security Group'
        });
        dltEcsSecurityGroup.addMetadata('cfn_nag', {
            rules_to_suppress: [{
                id: 'W40',
                reason: 'IpProtocol set to -1 (any) as ports are not known prior to running tests'
            }]
        });

        this.dltSecurityGroupId = dltEcsSecurityGroup.ref;

        new CfnSecurityGroupEgress(this, 'DLTSecGroupEgress', {
            cidrIp: props.securityGroupEgress,
            description: 'Allow tasks to call out to external resources',
            groupId: dltEcsSecurityGroup.ref,
            ipProtocol: '-1'
        });

        new CfnSecurityGroupIngress(this, 'DLTSecGroupIngress', {
            description: 'Allow tasks to communicate',
            fromPort: 50000,
            groupId: dltEcsSecurityGroup.ref,
            ipProtocol: 'tcp',
            sourceSecurityGroupId: dltEcsSecurityGroup.ref,
            toPort: 50000
        });
    }
}
