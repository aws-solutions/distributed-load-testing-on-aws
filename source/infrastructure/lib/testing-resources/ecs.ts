// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource } from "aws-cdk-lib";
import { CfnCluster, CfnTaskDefinition } from "aws-cdk-lib/aws-ecs";
import {
  Effect,
  ManagedPolicy,
  ServicePrincipal,
  Role,
  PolicyDocument,
  PolicyStatement,
  Policy,
} from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface ECSResourcesConstructProps {
  readonly cloudWatchLogsPolicy: Policy;
  // Container image
  readonly containerImage: string;
  // Fargate VPC ID
  readonly fargateVpcId: string;
  // Scenarios S3 bucket
  readonly scenariosS3Bucket: string;
  // IP CIDR for Fargate egress
  readonly securityGroupEgress: string;
  // Solution ID
  readonly solutionId: string;
}

/**
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class ECSResourcesConstruct extends Construct {
  public taskClusterName: string;
  public ecsCloudWatchLogGroup: LogGroup;
  public taskDefinitionArn: string;
  public taskExecutionRoleArn: string;
  public ecsSecurityGroupId: string;
  public taskDefinitionFamily: string | undefined;

  constructor(scope: Construct, id: string, props: ECSResourcesConstructProps) {
    super(scope, id);

    const dltTaskCluster = new CfnCluster(this, "DLTEcsCluster", {
      clusterName: Aws.STACK_NAME,
      clusterSettings: [{ name: "containerInsights", value: "enabled" }],
      tags: [
        {
          key: "SolutionId",
          value: props.solutionId,
        },
        {
          key: "CloudFormation Stack",
          value: Aws.STACK_NAME,
        },
      ],
    });

    this.taskClusterName = dltTaskCluster.ref;

    const scenariosBucketArn = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket).bucketArn;

    const dltTaskExecutionRole = new Role(this, "DLTTaskExecutionRole", {
      assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"),
      managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName("service-role/AmazonECSTaskExecutionRolePolicy")],
      inlinePolicies: {
        ScenariosS3Policy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:ListBucket"],
              resources: [scenariosBucketArn, `${scenariosBucketArn}/*`],
            }),
          ],
        }),
      },
    });
    dltTaskExecutionRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    this.taskExecutionRoleArn = dltTaskExecutionRole.roleArn;

    this.ecsCloudWatchLogGroup = new LogGroup(this, "DLTCloudWatchLogsGroup", {
      retention: RetentionDays.ONE_YEAR,
    });
    const dltLogsGroupResource = this.ecsCloudWatchLogGroup.node.defaultChild as CfnResource;
    dltLogsGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });

    const dltTaskDefinition = new CfnTaskDefinition(this, "DLTTaskDefinition", {
      cpu: "2048",
      memory: "4096",
      networkMode: "awsvpc",
      executionRoleArn: this.taskExecutionRoleArn,
      requiresCompatibilities: ["FARGATE"],
      taskRoleArn: this.taskExecutionRoleArn,
      family: `${Aws.STACK_NAME}-task-family`,
      containerDefinitions: [
        {
          essential: true,
          name: `${Aws.STACK_NAME}-load-tester`,
          image: props.containerImage,
          memory: 4096,
          logConfiguration: {
            logDriver: "awslogs",
            options: {
              "awslogs-group": this.ecsCloudWatchLogGroup.logGroupName,
              "awslogs-stream-prefix": "load-testing",
              "awslogs-region": `${Aws.REGION}`,
            },
          },
        },
      ],
    });

    this.taskDefinitionArn = dltTaskDefinition.ref;
    this.taskDefinitionFamily = dltTaskDefinition.family;

    const ecsSecurityGroup = new CfnSecurityGroup(this, "DLTEcsSecurityGroup", {
      vpcId: props.fargateVpcId,
      groupDescription: "DLTS Tasks Security Group",
    });
    ecsSecurityGroup.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W40",
          reason: "IpProtocol set to -1 (any) as ports are not known prior to running tests",
        },
      ],
    });

    this.ecsSecurityGroupId = ecsSecurityGroup.ref;

    new CfnSecurityGroupEgress(this, "DLTSecGroupEgress", {
      cidrIp: props.securityGroupEgress,
      description: "Allow tasks to call out to external resources",
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "-1",
    });

    new CfnSecurityGroupIngress(this, "DLTSecGroupIngress", {
      description: "Allow tasks to communicate",
      fromPort: 50000,
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "tcp",
      sourceSecurityGroupId: ecsSecurityGroup.ref,
      toPort: 50000,
    });
  }
}
