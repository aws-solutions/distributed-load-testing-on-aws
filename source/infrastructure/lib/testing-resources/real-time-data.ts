// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, CfnResource, Duration, Stack } from "aws-cdk-lib";
import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Effect, PolicyStatement, PolicyDocument, Role, ServicePrincipal, Policy } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { LogGroup, FilterPattern } from "aws-cdk-lib/aws-logs";
import { LambdaDestination } from "aws-cdk-lib/aws-logs-destinations";
import { Construct } from "constructs";

export interface RealTimeDataConstructProps {
  readonly cloudWatchLogsPolicy: Policy;
  readonly ecsCloudWatchLogGroup: LogGroup;
  readonly iotEndpoint: string;
  readonly mainRegion: string;
  /**
   * Solution config properties.
   * solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sourceCodeBucket: IBucket;
  readonly sourceCodePrefix: string;
}

export class RealTimeDataConstruct extends Construct {
  constructor(scope: Construct, id: string, props: RealTimeDataConstructProps) {
    super(scope, id);

    const realTimeDataPublisherRole = new Role(this, "realTimeDataPublisherRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        IoTPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Publish"],
              resources: [
                Stack.of(this).formatArn({
                  region: props.mainRegion,
                  service: "iot",
                  resource: "topic",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
          ],
        }),
      },
    });
    realTimeDataPublisherRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const realTimeDataPublisher = new LambdaFunction(this, "RealTimeDataPublisher", {
      description: "Real time data publisher",
      handler: "index.handler",
      role: realTimeDataPublisherRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/real-time-data-publisher.zip`),
      runtime: Runtime.NODEJS_18_X,
      timeout: Duration.seconds(180),
      environment: {
        MAIN_REGION: props.mainRegion,
        IOT_ENDPOINT: props.iotEndpoint,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
      },
    });

    const realTimeDataPublisherResource = realTimeDataPublisher.node.defaultChild as CfnResource;
    realTimeDataPublisherResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    const ecsCloudWatchSubscriptionFilter = props.ecsCloudWatchLogGroup.addSubscriptionFilter(
      "ECSLogSubscriptionFilter",
      {
        destination: new LambdaDestination(realTimeDataPublisher),
        filterPattern: FilterPattern.allTerms("INFO: Current:", "live=true"),
      }
    );
    const subscriptionFilterPermission = ecsCloudWatchSubscriptionFilter.node.findChild("CanInvokeLambda");
    if (subscriptionFilterPermission != null)
      ecsCloudWatchSubscriptionFilter.node.addDependency(subscriptionFilterPermission);
  }
}
