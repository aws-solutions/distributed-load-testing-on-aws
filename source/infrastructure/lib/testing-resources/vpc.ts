// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Tags } from "aws-cdk-lib";
import { IpAddresses, SubnetType, Vpc } from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";

/**
 * Distributed Load Testing on AWS VPC construct.
 * Creates a VPC for the Fargate test tasks.
 * Includes 2 public subnets across 2 availability zones.
 */
export class FargateVpcConstruct extends Construct {
  public vpc: Vpc;

  constructor(scope: Construct, id: string, vpcCidrBlock: string) {
    super(scope, id);

    const fargateVpc = new Vpc(this, "DLTFargateVpc", {
      ipAddresses: IpAddresses.cidr("192.168.0.0/16"),
      enableDnsHostnames: true,
      enableDnsSupport: true,
      subnetConfiguration: [
        {
          name: "DLTSubnetA",
          subnetType: SubnetType.PUBLIC,
          cidrMask: 20,
        },
        {
          name: "DLTSubnetB",
          subnetType: SubnetType.PUBLIC,
          cidrMask: 20,
        },
      ],
      maxAzs: 2,
      natGateways: 0,
    });

    fargateVpc.node.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W60",
          reason: "This VPC is used for the test runner Fargate tasks only, it does not require VPC flow logs.",
        },
      ],
    });

    this.vpc = fargateVpc;

    Tags.of(fargateVpc).add("Name", Aws.STACK_NAME);
  }
}
