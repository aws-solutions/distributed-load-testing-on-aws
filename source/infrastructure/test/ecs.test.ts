// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { ECSResourcesConstruct } from "../lib/testing-resources/ecs";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import { createTemplateWithoutS3Key } from "./snapshot_helpers";

test("DLT ECS Test", () => {
  const app = new App();
  process.env.PUBLIC_ECR_REGISTRY = "registry";
  process.env.PUBLIC_ECR_TAG = "tag";
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
      imageAssetsRepositoryName: process.env.PUBLIC_ECR_REGISTRY,
      dockerTagPrefix: process.env.PUBLIC_ECR_TAG,
    }),
  });
  const testPolicy = new Policy(stack, "TestPolicy", {
    statements: [
      new PolicyStatement({
        resources: ["*"],
        actions: ["cloudwatch:Get*"],
      }),
    ],
  });
  const vpc = new Vpc(stack, "TestVPC");
  const ecs = new ECSResourcesConstruct(stack, "TestECS", {
    fargateVpc: vpc,
    scenariosS3Bucket: "testscenariobucket",
    securityGroupEgress: "0.0.0.0/0",
    solutionId: "SO0062",
  });

  expect(createTemplateWithoutS3Key(stack)).toMatchSnapshot();
  Template.fromStack(stack).hasResourceProperties("AWS::ECS::Cluster", {
    ClusterSettings: [
      {
        Name: "containerInsights",
        Value: "enabled",
      },
    ],
  });
  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Role", {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: "sts:AssumeRole",
          Effect: "Allow",
          Principal: {
            Service: "ecs-tasks.amazonaws.com",
          },
        },
      ],
      Version: "2012-10-17",
    },
  });
  Template.fromStack(stack).hasResourceProperties("AWS::Logs::LogGroup", {
    RetentionInDays: 365,
  });
  Template.fromStack(stack).hasResourceProperties("AWS::ECS::TaskDefinition", {
    Cpu: "2048",
    Memory: "4096",
    NetworkMode: "awsvpc",
    RequiresCompatibilities: ["FARGATE"],
  });
  Template.fromStack(stack).resourceCountIs("AWS::EC2::SecurityGroup", 1);
  Template.fromStack(stack).hasResourceProperties("AWS::EC2::SecurityGroup", {
    SecurityGroupEgress: [
      {
        CidrIp: "0.0.0.0/0",
        IpProtocol: "-1",
      },
    ],
  });
  Template.fromStack(stack).hasResourceProperties("AWS::EC2::SecurityGroupIngress", {
    FromPort: 50000,
    ToPort: 50000,
  });
  expect(ecs.taskClusterName).toBeDefined();
  expect(ecs.ecsCloudWatchLogGroup).toBeDefined();
  expect(ecs.taskDefinitionArn).toBeDefined();
  expect(ecs.taskExecutionRoleArn).toBeDefined();
  expect(ecs.ecsSecurityGroupId).toBeDefined();
});
