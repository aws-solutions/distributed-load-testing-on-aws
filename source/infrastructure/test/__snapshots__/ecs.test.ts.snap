// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT ECS Test 1`] = `
{
  "Resources": {
    "TestECSDLTCloudWatchLogsGroup505959A6": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W84",
              "reason": "KMS encryption unnecessary for log group",
            },
          ],
        },
      },
      "Properties": {
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestECSDLTEcsCluster7C6F0F5D": {
      "Properties": {
        "ClusterName": {
          "Ref": "AWS::StackName",
        },
        "ClusterSettings": [
          {
            "Name": "containerInsights",
            "Value": "enabled",
          },
        ],
        "Tags": [
          {
            "Key": "CloudFormation Stack",
            "Value": {
              "Ref": "AWS::StackName",
            },
          },
          {
            "Key": "SolutionId",
            "Value": "SO0062",
          },
        ],
      },
      "Type": "AWS::ECS::Cluster",
    },
    "TestECSDLTEcsSecurityGroupFE5016DC": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W40",
              "reason": "IpProtocol set to -1 (any) as ports are not known prior to running tests",
            },
          ],
        },
      },
      "Properties": {
        "GroupDescription": "DLTS Tasks Security Group",
        "VpcId": "vpc-1a2b3c4d5e",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestECSDLTSecGroupEgressFC1E4757": {
      "Properties": {
        "CidrIp": "0.0.0.0/0",
        "Description": "Allow tasks to call out to external resources",
        "GroupId": {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "IpProtocol": "-1",
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "TestECSDLTSecGroupIngress78B83F3F": {
      "Properties": {
        "Description": "Allow tasks to communicate",
        "FromPort": 50000,
        "GroupId": {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Ref": "TestECSDLTEcsSecurityGroupFE5016DC",
        },
        "ToPort": 50000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "TestECSDLTTaskDefinitionB261038E": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": "testRepository/testImage:testTag",
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "TestECSDLTCloudWatchLogsGroup505959A6",
                },
                "awslogs-region": {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "load-testing",
              },
            },
            "Memory": 4096,
            "Name": {
              "Fn::Join": [
                "",
                [
                  {
                    "Ref": "AWS::StackName",
                  },
                  "-load-tester",
                ],
              ],
            },
          },
        ],
        "Cpu": "2048",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "TestECSDLTTaskExecutionRole104D057B",
            "Arn",
          ],
        },
        "Memory": "4096",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "TestECSDLTTaskExecutionRole104D057B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TestECSDLTTaskExecutionRole104D057B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "s3:HeadObject",
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:ListBucket",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::testscenariobucket",
                        ],
                      ],
                    },
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::testscenariobucket/*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ScenariosS3Policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestPolicyCC05E598": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": [
          {
            "Ref": "TestECSDLTTaskExecutionRole104D057B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
